### ISSUE TITLE
ImageNet model training for 1000 classes

### ISSUE THREAD

**Issue by tusharsoni08** [allanzelener/YAD2K#5]
Hi, I want to try this Keras version of YOLO for training on ImageNet model for 1000 classes, What changes do I have to make? Like for YOLOv1 they have mentioned about it here: https://pjreddie.com/darknet/yolov1/ in _Training YOLO_ section.
Any help for do that in this Keras version YOLO.

Thanks!

**Comment by allanzelener** [comment]
If you want to train on ImageNet classification from scratch you can use the Darknet19 model. (Note that I haven't thoroughly tested this myself.)

```python
from yad2k.models.yolo_darknet import darknet19
from keras.layers import Input, Activation

x = Input(shape=(224, 224, 3))
model = darknet19(x)
model.compile(optimizer='sgd', loss='categorical_crossentropy')
model.fit(X_train, Y_train)
```
You may want to check to see the optimizer settings in darknet19.cfg to get the same results. You'll also have to handle Imagenet preprocessing yourself.

**Comment by pengpaiSH** [comment]
@allanzelener I have checked ```darknet19.py``` source code and find out that the network aims at classification. The original paper says that yolo warms up by training a 19-layers darknet with ImageNet data. I am wondering if the classification network (darknet19) has already been built up, what changes shall we make in order to do the detection task?

**Comment by Tangzy7** [comment]
@allanzelener I want to train my own object detection task from scratch using your code. How should I input the labels(box information)?

**Comment by allanzelener** [comment]
@pengpaiSH @Tangzy7 The initial_loss_implementation branch has a preliminary implementation of the YOLO loss function needed for training. I've been busy this past week so I haven't had time to work on it but will probably get around to it this week.

@pengpaiSH If you train or convert a darknet19 model then you need to remove the last layer which gives the scores for the 1000 imagenet classes and add the layers given by the `yolo_head` function in `keras_yolo.py`.

**Comment by allanzelener** [comment]
I've now merged the loss implementation branch into master and have added a sample training script called `train_overfit.py` which overfits a YOLO_v2 model to a single image.

This script can be adapted to a full training script but personally I don't have time to replicate the full training procedure in the paper. If anyone is willing to expand the training script and confirm that the model can be trained on Pascal VOC/MS COCO then I'm sure that contribution would be appreciated.

**Comment by Tangzy7** [comment]
Thanks toyourcontribution. I am working on training yourscript on VOC.I just met a problem. I am sure that you know what I am doing.


The shape of the box tensor is (None, None ,5). The first’None’ should be the size of batch. The 2nd’None’ should be the number of boxes. So, how to make a numpy array as input which has twounknown length dimensions.


原始邮件
发件人:allanzelenernotifications@github.com
收件人:allanzelener/YAD2KYAD2K@noreply.github.com
抄送:Tangzy7tangzy048@163.com; Mentionmention@noreply.github.com
发送时间:2017年4月18日(周二) 19:46
主题:Re: [allanzelener/YAD2K] ImageNet model training for 1000 classes(#5)


I've now merged the loss implementation branch into master and have added a sample training script called train_overfit.py which overfits a YOLO_v2 model to a single image.
This script can be adapted to a full training script but personally I don't have time to replicate the full training procedure in the paper. If anyone is willing to expand the training script and confirm that the model can be trained on Pascal VOC/MS COCO then I'm sure that contribution would be appreciated.

**Comment by allanzelener** [comment]
@Tangzy7 Good question! The easy answer is to use batch_size=1 so you don't have to worry about it.

The more general solution is to zero pad the input boxes so all the examples in a batch have a number of boxes equal to the number of boxes in the example with the most boxes. Or just use a large fixed number of boxes like 10. The only thing is I'm not sure if the loss function will be computed correctly in this case, you may also need a mask for the padding boxes so they're not included in the loss.

**Comment by Tangzy7** [comment]
@allanzelener Good idea. I will zero pad the boxes to 100. I think the loss function is still correct in this case, because it seems that the function yolo_loss’s input parameter, true_boxes, has noinfluence on the total loss?
原始邮件
发件人:allanzelenernotifications@github.com
收件人:allanzelener/YAD2KYAD2K@noreply.github.com
抄送:Tangzy7tangzy048@163.com; Mentionmention@noreply.github.com
发送时间:2017年4月20日(周四) 15:29
主题:Re: [allanzelener/YAD2K] ImageNet model training for 1000 classes(#5)


@Tangzy7 Good question! The easy answer is to use batch_size=1 so you don't have to worry about it.
The more general solution is to zero pad the input boxes so all the examples in a batch have a number of boxes equal to the number of boxes in the example with the most boxes. Or just use a large fixed number of boxes like 10. The only thing is I'm not sure if the loss function will be computed correctly in this case, you may also need a mask for the padding boxes so they're not included in the loss.

**Comment by allanzelener** [comment]
true_boxes does have an influence on the total loss. It's used to determine which "detectors" have found an object by computing the IOU (intersection area over union area) of each detector's predicted box with each ground truth box and checking if it's greater than 0.6.

But since the intersection with the zeros box is zero the IOU will be 0 and no detector should activate for the zeros box so I think you're right that the loss function will work as is with zero padding for the true_boxes.

**Comment by AceCoooool** [comment]
Is the darknet loss the (7x7) avgpool layer? (before softmax)

**Comment by allanzelener** [comment]
@AceCoooool No, the loss function is not included in the model definition of `darknet19`. For Imagenet classification you can just use the `categorical_crossentropy` loss as shown in the second comment of this thread. (If you're instead referring to the YOLO loss function then that's contained in `keras_yolo.py` as `yolo_loss`. I'd recommend looking at the original YOLO paper to understand what that's doing since the code may be difficult to read.)

**Comment by AceCoooool** [comment]
thank you very much
